src/System.Management.Automation/engine/BytePipe.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#nullable enable
+
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+
+namespace System.Management.Automation;
+
+internal interface IStreamSource
+{
+    Stream GetStream();
+}
+
+internal sealed class StdOutStreamSource : IStreamSource
+{
+    private readonly Process _process;
+
+    public StdOutStreamSource(Process process) => _process = process;
+
+    public Stream GetStream() => _process.StandardOutput.BaseStream;
+}
+
+internal abstract class BytePipe : IStreamSource
+{
+    private readonly object _syncObject = new();
+
+    private int _refCount;
+
+    public abstract Stream GetStream();
+
+    internal AsyncByteStreamDrainer Bind(IStreamSource sourceStream)
+    {
+        Debug.Assert(sourceStream is not null);
+        Interlocked.Increment(ref _refCount);
+        try
+        {
+            return new AsyncByteStreamDrainer(
+                sourceStream,
+                (bytes, _) =>
+                {
+                    lock (_syncObject)
+                    {
+                        GetStream().Write(bytes);
+                    }
```

Comment:
The lock might not be needed here. Originally this was due to the drainers for `stdout` and `stderr` writing to the same stream when redirected, but now we revert to string reading for that.